pytorch_lightning/utilities/fetching.py

```diff
@@ -228,44 +228,33 @@ def wait(self) -> None:
         """Hook to override to indicate the `DataFetcher` to wait for an event."""
 
     def prefetching(self) -> None:
+        iterator = self.dataloader_iter
         for _ in range(self.prefetch_batches):
             try:
-                self._fetch_next_batch()
+                self._fetch_next_batch(iterator)
             except StopIteration:
                 break
 
-    def fetching_function(self) -> Optional[Tuple[Any, bool]]:
-        if self.done:
-            while self.batches:
-                return self._get_queued_batch()
-            raise StopIteration
+    def fetching_function(self) -> Tuple[Any, bool]:
+        if self.batches:
+            batch = self.batches.pop(0)
         else:
+            # empty iterator, no prefetching done
+            raise StopIteration
```

Comment:
This means, you won't get a batch here at all right? If no pre-fetching was done, we should still wait for a batch to be normally loaded, no? How can this work if you early-exit the function here? The `__iter__` imoplementation does not catch this and therefore it would result in end of iterating at all or am I missing something?

Reply:
pre-fetching=0 is not supported. See the original implementation before I started refactoring: https://github.com/PyTorchLightning/pytorch-lightning/blob/0066ff012962ffa920e9bbd750f4edbea05d4505/pytorch_lightning/utilities/fetching.py#L65-L71

This should be explicitly checked for, but needs further refactoring of the inheritance structure.

The only class that uses `prefetch_batches=0` is `DataLoaderIterDataFetcher` but it overrides `fetching_function`.